server/render/index.js

```diff
@@ -226,3 +199,52 @@ export function serverRenderError( err, req, res, next ) {
 
 	next();
 }
+
+/**
+ * Middlewares may decide to override (or add, or remove) the validators defined as defaults.
+ *
+ * Adds SSRValidators to the context, an object in which property names are idnetifiers and the values are callables
+ * that shouls return booleans. If one of the validators fails, then SSR is considered disabled. All methdos would
+ * receive the context on call.
+ *
+ * @param {object}   context  The entire request context
+ * @param {function} next     As all middlewares, will just add next
+ */
+export function initShouldSSR( context, next ) {
+	context.SSRValidators = {
+		config: () => config.isEnabled( 'server-side-rendering' ),
+		section: () => isSectionIsomorphic( context.store.getState() ),
+		user: () => ! context.user, // logged out
+		locale: () => isDefaultLocale( context.lang ),
+		layout: () => context.layout,
+		query: () => Object.keys( context.query ).length === 0, // no query args
+	};
+
+	next();
+}
+
+/**
+ * Checks if SSR is usable for the current context.
+ *
+ * @see also initShouldSSR, but consider the fact that middlwares may override the defaults.
+ *
+ * @param {object}   context The currently built context
+ *
+ * @return {boolean} If all conditions in the context.SSRValidators hashmap are met
+ */
+function shouldSSR( context ) {
+	if ( ! context.SSRValidators ) {
+		// because the section is not isomorphic, the middleware initShouldSSR is not called)
+		return false;
+	}
+
+	for ( const validatorKey in context.SSRValidators ) {
+		const validatorFunction = context.SSRValidators[ validatorKey ];
+		if ( ! validatorFunction( context ) ) {
+			debug( 'SSR Disabled due to the callable check on ' + validatorKey );
+			return false;
+		}
+	}
```

Comment:
It would be nice to throw in development, but provide some safety in production with a function check

```js
if ( 'function' !== typeof validatorFunction ) {
  if ( 'production' === process.env.NODE_ENV ) {
    throw new TypeError( 'Terrible things happen!' );
  }
  return false;
}
```